sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -387,6 +381,14 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Can't prune the columns on LeafNode
     case p @ Project(_, l: LeafNode) => p
 
+    // Prune windowExpressions and child of Window
+    case p @ Project(_, w: Window) if (w.outputSet -- p.references).nonEmpty =>
+      val newWindowExprs = w.windowExpressions.filter(p.references.contains)
+      val newGrandChild = prunedChild(w.child, w.references ++ p.references)
```

Comment:
`w.references ++ p.references` are not the attributes we need from child right? First, we may filter out some window expressions and `w.references` is out of date. Second, `w.child.output` will never contains `p.reference`, as `Window` operator only produce window function results.